src/main/java/edu/ucsb/cs156/courses/services/UCSBCurriculumService.java

```diff
@@ -166,4 +166,41 @@ public String getSectionJSON(String quarter, String enrollCode) {
         logger.info("json: {} contentType: {} statusCode: {}",retVal,contentType,statusCode);
         return retVal;
     }
-}
\ No newline at end of file
+
+  public static final String FINAL_ENDPOINT = "https://api.ucsb.edu/academics/curriculums/v1/finals";
+
+  public String getFinalJSON(String quarter, String enrollCode) {
+
+        HttpHeaders headers = new HttpHeaders();
+        headers.setAccept(Arrays.asList(MediaType.APPLICATION_JSON));
+        headers.setContentType(MediaType.APPLICATION_JSON);
+        headers.set("ucsb-api-version", "1.0");
+        headers.set("ucsb-api-key", this.apiKey);
+
+        HttpEntity<String> entity = new HttpEntity<>("body", headers);
+
+        String params = String.format(
+            "?quarter=%s&enrollCode=%s", quarter, enrollCode);
+        String url = FINAL_ENDPOINT + params;
+
+        logger.info("url=" + url);
+
+        String retVal = "";
+        MediaType contentType=null;
+        HttpStatus statusCode=null;
+        try {
+            ResponseEntity<String> re = restTemplate.exchange(url, HttpMethod.GET, entity, String.class);
+            contentType = re.getHeaders().getContentType();
+            statusCode = re.getStatusCode();
+            retVal = re.getBody();
+        } catch (HttpClientErrorException e) {
+            retVal = "{\"error\": \"401: Unauthorized\"}";
+        }
+        if (retVal.contains("null")){
+            retVal = "{\"error\": \"404 (Not Found): Enroll code does not exist!\"}";
+        }
```

Comment:
I don't think this check is valid here. The code correctly enters this case when an enroll code does not exist, but this case is also being hit if the enroll code exists but the course doesn't have a final time, which is incorrect behavior. Try the following values:

Quarter: 20222, Enroll Code: 60673
CMPSC 292C

Quarter: 20222, Enroll Code: 08912
CMPSC 502

Quarter: 20201, Enroll Code: 12567
ECE 1A

Reply:
addressed by changing error message to indicate no final exam info was found.